### PR TITLE
Merge CUDA Streams and Events

### PR DESCRIPTION
**Pull request by nikithamalgifb**
Summary:
-----------
- Updates current_stream and default stream API's to take `optional[device]` argument
- Adds parsing logic to replace `torch.cuda.Stream` and `torch.cuda.Event` -> `torch.classes.cuda.Stream` and `torch.classes.cuda.Event` for JIT
- Merges StreamContext manager for both Eager and JIT.

Test:
------
Run JIT tests:
python test/test_jit.py -v TestCUDA

Run eager tests:
python test/test_cuda.py -v TestCuda

### CODE REVIEW

File: torch/cuda/__init__.py
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:350]: This index variable is required to check if there is no CUDA device is available. Also, it is later referred in `__exit__` to check the same. If we want to eliminate `self.idx` completely, then we would need to call this API `_get_device_index` in `__enter__` and `__exit__`

Comment (SplitInfinity) [reply torch/cuda/__init__.py:350]: The old implementation of this context manager didn't have this check. It only checks if `stream` is `None` and exits early in that case.

Comment (nikithamalgifb) [reply torch/cuda/__init__.py:350]: Yes, it was added in case the CUDA device isn't available.

Comment (SplitInfinity) [reply torch/cuda/__init__.py:350]: How come the old implementation did not have code like this? Is it a bug in their implementation?

If there is no CUDA device available, the stream must be `None`, right?

Comment (nikithamalgifb) [reply torch/cuda/__init__.py:350]: How come the old implementation did not have code like this? Is it a bug in their implementation?
>> Maybe.Because if the CUDA device becomes unavailable inside `__enter__` , then `stream` is still not `None`(as it was set in `__init__`  and we don't return.

Comment (SplitInfinity) [reply torch/cuda/__init__.py:350]: Ok, so it is an issue introduced by this implementation because `__init__` and `__enter__` are two separate functions. I think I understand now.

But under what circumstances could a device be available in `__init__` and not in `__enter__`?

Comment (nikithamalgifb) [reply torch/cuda/__init__.py:350]: The only scenario that I can think of is when the connection to device is lost in `__enter__`. Although the window is very tiny, there are chances of this still happening.

File: torch/jit/frontend.py
Comment (SplitInfinity) [inline torch/jit/frontend.py:235]: Is it possible to add this logic in the type resolver/related type resolution code? Something similar to how we do type resolution for `ScriptClass`:

https://github.com/pytorch/pytorch/blob/65087dd1d4c1e503281cd676e5563e9e64b2a926/torch/csrc/jit/python/python_sugared_value.cpp#L1161-L1165

**Comment by SplitInfinity** [comment]
Have you also run the performance benchmark from last time? I think there is a chat group in which someone shared the command for the benchmark.

File: torch/jit/frontend.py
Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: I haven't tried this. Let me try it out.

**Comment by nikithamalgifb** [comment]
> Have you also run the performance benchmark from last time? I think there is a chat group in which someone shared the command for the benchmark.

Do we have separate tests for performance apart from the pytorch/benchmark models?

**Comment by SplitInfinity** [comment]
> > Have you also run the performance benchmark from last time? I think there is a chat group in which someone shared the command for the benchmark.
> 
> Do we have separate tests for performance apart from the pytorch/benchmark models?

Not that I know of. I'm referring specifically to an internal test from an internal CUDA user that @ngimel introduced us to.

**Comment by SplitInfinity** [comment]
@ezyang kindly let us know what you think. This implementation is cleaner than our last attempt.

**Comment by ngimel** [comment]
Can you run a benchmark counting python ops, similar to how it's done in #50431? You'd need to create some microbenchmarks stressing just stream context manager or streams operation.

File: test/jit/test_cuda.py
Comment (ezyang) [inline test/jit/test_cuda.py:102]: This doesn't look very backwards compatible

**Review by SplitInfinity — COMMENTED** [review]
Two small suggestions for handling the BC issue raised by @ezyang (although we know that most likely nobody is using this feature yet...).

File: test/jit/test_cuda.py
Comment (SplitInfinity) [reply test/jit/test_cuda.py:102]: Nobody is using this feature yet (they are waiting for this diff to land first) so I don't think BC is a problem. However, the eager mode `current_stream` can accept `device`, `int`, or `str`, so we need to do that in order to get API parity. I initially thought it would not be possible due to the lack of a `Union` type in JIT, so I thought supporting `device` arguments would make the most sense.

However, this API is desugared and not compiled, so I think multiple JIT operators can be registered for it with different input types. I commented elsewhere in the PR on how to do this.

Comment (ezyang) [reply test/jit/test_cuda.py:102]: Oh, so just to make sure I understand you correctly, both int and device work in eager mode, but in the current version of this diff, only device works inside torch.jit.script? I agree this is less bad then.

Comment (SplitInfinity) [reply test/jit/test_cuda.py:102]: Yes.

File: torch/_utils.py
Comment (ezyang) [inline torch/_utils.py:466]: Underscore me?

Comment (nikithamalgifb) [reply torch/_utils.py:466]: The eager version of the API has an `_` -> `_get_current_device_index`. So couldn't use. Do you suggest changing the name instead?

Comment (ngimel) [reply torch/_utils.py:466]: Can you unify eager and jit implementations for get_current_device_index? It looks like they are doing the same thing, spelled slightly differently, like, using `torch.cuda.is_available()` vs `torch.cuda.device_count()`

Comment (nikithamalgifb) [reply torch/_utils.py:466]: `_get_current_device_index` currently uses `lambda` expression which is not supported in JIT. That's the reason for introducing a separate API for JIT

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp]: Done

File: torch/cuda/_utils.py
Comment (ezyang) [inline torch/cuda/_utils.py:3]: don't forget to fix the lint error

File: torch/_utils.py
Comment (ezyang) [reply torch/_utils.py:466]: The implementation is pretty simple:

```
def _get_device_attr(get_member):
    device_type = _get_available_device_type()
    if device_type and device_type.lower() == "cuda":
        return get_member(torch.cuda)
    # add more available device types here
    return None


def _get_current_device_index():
    # current device index
    return _get_device_attr(lambda m: m.current_device())
```

TBH, you should just inline the lambda and trace it directly

Comment (ngimel) [reply torch/_utils.py:466]: If perf of eager and jit implementation is the same, then eager can use your jit implementation. You can use instruction counts to measure perf (`collect_callgrind` https://pytorch.org/tutorials/recipes/recipes/benchmark.html), timer may be too noisy.

File: torch/cuda/__init__.py
Comment (ezyang) [inline torch/cuda/__init__.py:334]: botched merge

File: torch/cuda/__init__.py
Comment (ezyang) [inline torch/cuda/__init__.py:458]: botched merge

File: torch/jit/frontend.py
Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: `torch.cuda.Stream` calls the eager mode `Stream` if we do not replace it in `frontend.py`. So this won't be possible to defer the type resolution later in  `python_sugared_value.cpp`.

File: torch/cuda/_utils.py
Comment (nikithamalgifb) [reply torch/cuda/_utils.py]: Yup. Fixed this

File: torch/cuda/__init__.py
Comment (nikithamalgifb) [reply torch/cuda/__init__.py]: Missed this. Fixed now

File: torch/cuda/__init__.py
Comment (nikithamalgifb) [reply torch/cuda/__init__.py]: Fixed now.

File: torch/jit/frontend.py
Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: Hmm, but I think that happens because `CUDAModuleValue::attr` uses `PythonValue::attr` when the attribute is not one of the functions we have a custom operator for, right? 

https://github.com/pytorch/pytorch/blob/b936abd840d8fa3122d3a6166c0d7344d8352901/torch/csrc/jit/python/python_sugared_value.cpp#L235-L253

Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: Yeah, so that gets resolved as object of class torch.cuda.streams.Stream 

`obj name::<class 'torch.cuda.streams.Stream'>
I enabled a few debug prints , so a call to `torch.cuda.Stream()` results in these values:
 
```
obj name::<module 'torch' from '/data/users/nikithamalgi/pytorch/torch/__init__.py'>
name::::torch
obj name::<module 'torch.cuda' from '/data/users/nikithamalgi/pytorch/torch/cuda/__init__.py'>
name::::torch.cuda
field:::Stream
**obj name::<class 'torch.cuda.streams.Stream'>**
```
Here `torch.cuda` results in `CUDAModuleValue::attr` while `torch.classes.cuda` results in `PythonValue::getattr`.

Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: So we should be able to return an instance of `ClassValue` corresponding to the custom class that we registered for `Stream` when TorchScript tries to use `torch.cuda.Stream`, right?

This doesn't handle resolving type annotations, but it's part of the way there, right?

Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: I tried returning `ClassValue` corresponding to custom class and I ended up in compilation errors.

Here is what I tried:
```
return std::make_shared<PythonClassValue>(
        py::cast<torch::jit::CUDAStream>(member));
./torch/csrc/jit/python/python_sugared_value.h:300:26: note:   no known conversion for argument 1 from ‘torch::jit::CUDAStream’ to ‘const torch::jit::PythonClassValue&’
```
When `torch.classes.cuda.Stream` is used, it gets resolved as `torch._C.ScriptClass`.

**Comment by nikithamalgifb** [comment]
> Can you run a benchmark counting python ops, similar to how it's done in #50431? You'd need to create some microbenchmarks stressing just stream context manager or streams operation.

After running the counting python ops, here are some numbers:

You can find these microbenchmark tests in `test_cuda.py` for reference. 

Test                                                            c.ops.       c.torches.    c.ops / c.torches
===========================================================
test_stream_context (Master)                  1761               4                     440.25
test_stream_context (merge stream)      2837              4                     709.25
test_multiple_stream (Master)                 3232              8                     404.0
test_multiple_stream (merge stream)     5716               8                     714.5

File: torch/jit/frontend.py
Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: Oh, but that's a `PythonClassValue`, and we don't have a Python object in this case. What about a regular `ClassValue`?

Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: Yes regular `ClassValue` also resulted in the same error :
```
../torch/csrc/jit/frontend/sugared_value.h:335:18: note:   no known conversion for argument 1 from ‘torch::jit::CUDAStream’ to ‘const torch::jit::ClassValue&’
```

Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: The constructor for `ClassValue` accepts a `ClassType`, not a `jit::CUDAStream`.

Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: More specifically, it is a `ClassTypePtr` (i.e. `std::shared_ptr<ClassType>`), not a raw `ClassType` instance.

Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: OK `ClassTypePtr` compiled, but I ran into RuntimeError:

```
RuntimeError: Unable to cast Python instance of type <class 'type'> to C++ type 'std::shared_ptr<c10::ClassType>'
```

Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: It seems like you tried to cast the Python object representing `torch.cuda.Stream` into a `ClassTypePtr`? You should look up the `ClassType` for `jit::Stream` in the compilation unit and use that.

Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: As suggested, tried the above. But I'm still running into errors.
Here is how I did it:
```
    auto pyCu = get_python_cu();
    auto qualname = c10::QualifiedName("torch::jit::CUDAStream");
    auto type = pyCu->get_type(qualname);
    return std::make_shared<ClassValue>(type);
```

Also, `torch.classes.cuda.Stream` resolves to  `torch._C.ScriptClass`. I also tried replacing `torch::jit::CUDAStream` with `torch.classes.cuda.Stream`, that did not work either.

Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: Isn't the fully qualified name `__torch__.torch.classes.cuda.Stream`?

Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: Yes, `__torch__.torch.classes.cuda.Stream` that did not work either. I tried making it a PythonClassValue as well, didn't work.

File: torch/cuda/__init__.py
Comment (ezyang) [inline torch/cuda/__init__.py:344]: You don't need to initialize these if stream is None, because the context manager is a noop in that case.

File: torch/jit/frontend.py
Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: Can you dump all of the registered classes in the compilation unit? Maybe we're missing something.

When you try `__torch__.torch.classes.cuda.Stream`, is the class type successfully found? Is it the same error as before?

File: torch/cuda/__init__.py
Comment (ezyang) [reply torch/cuda/__init__.py:350]: By this reasoning, this code could also be wrong, because you could test this in line 347, and then immediately afterwards we could lose connection to device.

The more realistic approach is to assume (like most code in PyTorch does) that the set of available CUDA devices stays fixed for the entirety of the PyTorch process. Then this code is unnecessary

File: torch/cuda/__init__.py
Comment (ezyang) [inline torch/cuda/__init__.py:353]: At least mypy is able to handle type refinement of class members. Small test case:

```
ezyang-mbp:~ ezyang$ cat a.py
from typing import Optional
class A:
    b: Optional[int]
    def f(self) -> None:
        assert self.b is not None
        reveal_type(self.b)
ezyang-mbp:~ ezyang$ mypy a.py
a.py:6: error: Revealed type is 'builtins.int'
```

Is TorchScript unable to do this?

**Comment by ezyang** [comment]
There's some weirdness in the nitty details of the new implementation, but broadly speaking it looks a lot better now. Check comments for specific remarks.

File: torch/jit/frontend.py
Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: Registeration is the issue here. `__torch__.torch.classes.cuda.Stream` was not registered.  I was able to resolve this, but I am still running into errors like: 
```
_jit_script_compile qualname --__torch__.fn
find_function:__torch__.fn
field:::Stream
Registered classes for :::__torch__.torch.classes.cuda.Stream
{__torch__.torch.classes.cuda.Stream: 0}
it->second0x559dee724130
Traceback (most recent call last):
  File "repro.py", line 4, in <module>
    def fn():
  File "/data/users/nikithamalgi/pytorch/torch/jit/_script.py", line 994, in script
    fn = torch._C._jit_script_compile(
RuntimeError: Unable to cast Python instance of type <class 'type'> to C++ type 'torch::jit::ScriptClass'
```
I'm not sure why class results to `type` here. 

The code is:
```
    auto pyCu = get_python_cu();
    auto qualname = c10::QualifiedName("__torch__.torch.classes.cuda.Stream");
    auto class_type = ClassType::create(qualname, pyCu);
    pyCu->register_type(class_type);
    return std::make_shared<ClassValue>(class_type);
```

File: torch/cuda/__init__.py
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:353]: Yes, that's right.

File: torch/cuda/__init__.py
Comment (nikithamalgifb) [reply torch/cuda/__init__.py:344]: Agreed. TorchScript complains if we do something like this:
```
if self.stream is not None:
        self.src_prev_stream = torch.cuda.default_stream(None)
        self.dst_prev_stream = torch.cuda.default_stream(None)

RuntimeError: 
First assignment cannot be in a control-flow block. Initialize the field at the top level first:
```

File: torch/jit/frontend.py
Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: The `ClassType` should already exist, so you should not create it like that. 

Can you add some debug code to `CompilationUnit` to print out all the types so that we can determine if the type name is being spelling correctly? TorchBind uses static registration, so the type should already be registered.

Comment (nikithamalgifb) [reply torch/jit/frontend.py:235]: Well the register class types return empty.

Here is a log

```
_jit_script_compile qualname --__torch__.fn
find_function:__torch__.fn
All registered class types   # This did not print anything
Register Function - __torch__.fn
field:::Stream
```

Comment (SplitInfinity) [reply torch/jit/frontend.py:235]: > TorchBind uses static registration, so the type should already be registered.

I was right, but not about *where* the types are stored. Can you try using `getCustomClass()` instead of `CompiltationUnit::get_type`.

File: torch/csrc/jit/frontend/script_type_parser.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/frontend/script_type_parser.cpp:257]: What's the error you are getting due to this?

One thing to check is whether this code above is resolving the type for these classes before they have a chance to execute:

```
      if (auto typePtr = resolver_->resolveType(type_name, expr.range())) {
        return typePtr;
      }
```

Comment (nikithamalgifb) [reply torch/csrc/jit/frontend/script_type_parser.cpp:257]: Yeah, it is not able to resolve the type inside `stream`. It resolves to `torch.cuda.streams.Stream`, the eager mode `Stream`.

Comment (SplitInfinity) [reply torch/csrc/jit/frontend/script_type_parser.cpp:257]: I would check if the code I quoted above is resolving the type before the code you added is able to run.

File: torch/cuda/__init__.py
Comment (nikithamalgifb) [reply torch/cuda/__init__.py]: Sounds good. Fixed now

File: torch/jit/frontend.py
Comment (nikithamalgifb) [reply torch/jit/frontend.py]: Fixed now.

File: torch/csrc/jit/frontend/script_type_parser.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/frontend/script_type_parser.cpp:261]: I hope moving this doesn't break any tests...

File: torch/csrc/jit/python/python_sugared_value.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/python/python_sugared_value.cpp:252]:
```suggestion
  if (field == "Stream" || field == "Event") {
    auto class_type = getCustomClass("__torch__.torch.classes.cuda." + field);
    return std::make_shared<ClassValue>(class_type);
  }
  
  py::object member = getattr(loc, field);

```

Comment (SplitInfinity) [inline torch/csrc/jit/python/python_sugared_value.cpp:252]: This should moved above the `getattr` above.

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/runtime/register_cuda_ops.cpp:26]: Why is this `c10::cuda::current_device()` here but `-1` below?

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (SplitInfinity) [inline torch/csrc/jit/runtime/register_cuda_ops.cpp:49]: Same here.

**Comment by ezyang** [comment]
Giving a basic LGTM from my end; the new manager is worse but it seems clear that these are mostly TorchScript induced so it seems like an OK tradeoff to make.

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:26]: This is due to clang-formatting.

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:49]: This is due to clang-formatting.

File: torch/csrc/jit/frontend/script_type_parser.cpp
Comment (nikithamalgifb) [reply torch/csrc/jit/frontend/script_type_parser.cpp:261]: All the tests have passed. The failures seen below are due to issues in reporting the results.

File: torch/csrc/jit/python/python_sugared_value.cpp
Comment (nikithamalgifb) [reply torch/csrc/jit/python/python_sugared_value.cpp:251]: Done

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:26]: `clang-format` changed the content of the code?

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (SplitInfinity) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:49]: `clang-format` changed the content of the code?

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:26]: No, sorry. I misread your question. No particular reason. `c10::cuda::current_device()`  returns the current device index. If we pass -1 to `getCurrentCUDAStream`, it uses the the current device to get the `stream`. Let me make it more consistent.

Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:26]: Fixed now.

File: torch/csrc/jit/runtime/register_cuda_ops.cpp
Comment (nikithamalgifb) [reply torch/csrc/jit/runtime/register_cuda_ops.cpp:50]: Fixed now.

**Comment by nikithamalgifb** [comment]
> > Can you run a benchmark counting python ops, similar to how it's done in #50431? You'd need to create some microbenchmarks stressing just stream context manager or streams operation.
> 
> @ngimel :
> After running the counting python ops, here are some numbers:
> 
> You can find these microbenchmark tests in `test_cuda.py` for reference.
> 
> # Test | ops | torches | ops / torches
> test_stream_context (Master) | 1761 | 4 | 440.25
> test_stream_context (merge stream) | 2837 | 4 | 709.25
> test_multiple_stream (Master) | 3232 | 8 | 404.0
> test_multiple_stream (merge stream) | 5716 | 8 | 714.5
> 
> PS: This display is a little messed up. I couldn't insert a table here.

@ngimel : Any thoughts about these numbers?

**Comment by SplitInfinity** [comment]
I have only a superficial understanding of what these numbers mean, but an almost 2x regression seems bad. I wonder if there is anything we can do to understand why this increase has occurred and how to fix it?

**Comment by SplitInfinity** [comment]
Just for the record, I have no more feedback on this. I am also waiting for ngimel's opinion before accepting.

**Comment by nikithamalgifb** [comment]
> > Can you run a benchmark counting python ops, similar to how it's done in #50431? You'd need to create some microbenchmarks stressing just stream context manager or streams operation.
> 
> @ngimel :
> After running the counting python ops, here are some numbers:
> 
> You can find these microbenchmark tests in `test_cuda.py` for reference.
> 
> # Test | ops | torches | ops / torches
> test_stream_context (Master) | 1761 | 4 | 440.25
> test_stream_context (merge stream) | 2837 | 4 | 709.25
> test_multiple_stream (Master) | 3232 | 8 | 404.0
> test_multiple_stream (merge stream) | 5716 | 8 | 714.5
> 
> PS: This display is a little messed up. I couldn't insert a table here.



> I have only a superficial understanding of what these numbers mean, but an almost 2x regression seems bad. I wonder if there is anything we can do to understand why this increase has occurred and how to fix it?

Did some investigation, looks like most of it is coming from the new context manager + combination of replacing torch._C.* APIs to torch.cuda.*  :(. I know this is not ideal, I am not sure what else can we do to reduce the counts.

**Comment by nikithamalgifb** [comment]
> Just for the record, I have no more feedback on this. I am also waiting for ngimel's opinion before accepting.

I have pinged @ngimel offline to provide her thoughts here, so that we can proceed further.

**Review by SplitInfinity — APPROVED** [review]
Make sure to address `flake8-3` errors before merging.

**Pull request closed without merging**
